### PR TITLE
Allocate dQ, dK, and dV as a catted tensor to save a downstream cat in nvFuser.

### PR DESCRIPTION
**Pull request by wujingyue**
For https://github.com/NVIDIA/Fuser/issues/1768. 

The description on the added compile option explains what this optimization does.

This optimization is disabled by default for now. I'll try to enable it by default or even always after #35 is merged and bookend is [disabled by default](https://github.com/Lightning-AI/lit-thunder-LEGACY/issues/1562).

### CODE REVIEW

File: thunder/tests/test_cudnn_executor.py [hunk 1]
```diff
@@ -209,25 +206,25 @@ def test_vjp_correctness_sdpa_cudnnex_manual(op, device, dtype, executor, comp):
...
-            executors_list=executor.executors_list() + [cudnn_ex],
+            executors_list=[cudnn_ex],
+            cudnn_sdpa_bwd_may_cat_grad_qkv=may_cat_grad_qkv,
         )
 
         actual_out, actual_grad = cfoo(filtered_args, (v,))
 
-        comp(actual_out, expect_out, atol=1e-2, rtol=1e-2)
-
+        torch.testing.assert_close(actual_out, expect_out, atol=1e-2, rtol=1e-2)
```
Comment (jjsjann123) [inline thunder/tests/test_cudnn_executor.py:227]: wondering if tolerance should be part of the opinfo? Not really something that we should do in this PR.

**Review by jjsjann123 — APPROVED** [review]
:clap:

File: thunder/tests/test_cudnn_executor.py [hunk 2]
```diff
@@ -186,15 +188,10 @@ def test_cudnn_vs_torch_consistency(op, device, dtype, *_):
             return result
 
 
-# NOTE Scaled_Dot_Product_Efficient_Attention_Backward does not support fp64 dtypes
-# RuntimeError: Only fp32, half & bf16 supported at the moment
-@ops(
```
Comment (wujingyue) [inline thunder/tests/test_cudnn_executor.py:191]: I didn't manage to add an extra parameter (e.g. `may_cat_grad_qkv`) to the `ops` instantiation. So I reworked (and happened to simplify) the parameterization part of the test.

File: thunder/tests/test_cudnn_executor.py (hunk 1 above)
Comment (wujingyue) [reply thunder/tests/test_cudnn_executor.py:227]: It's already allowed to be, e.g., https://github.com/Lightning-AI/lightning-thunder/blob/c67c81428000da9fe26e5c62f750ede9a87ed361/thunder/tests/opinfos.py#L6409. Figuring out the right threshold for cudnn sdpa, however, is real art to me.

**Review by t-vi — APPROVED** [review]
Per @jjsjann123 review. Thank you @wujingyue @jjsjann123 .

**Comment by wujingyue** [comment]
Hey @vedaanta-nvidia, are you interested in reviewing this before I merge?

File: thunder/tests/test_cudnn_executor.py (hunk 1 above)
Comment (vedaanta-nvidia) [reply thunder/tests/test_cudnn_executor.py:227]: True.

bf16 tolerances used for internal cudnn testing are also always higher than fp16.
Right now. sdpa tests in cudnn testing do:
`atol=2e-2 if input_type != torch.bfloat16 else 7e-2, rtol=2e-2`

Plus, it is known that there is slightly large mismatches between flash attention based kernels compared to pure math kernels. So depending on what reference is being run above, there are large swings in absolute difference.

Comment (wujingyue) [reply thunder/tests/test_cudnn_executor.py:227]: Got it. Do you request a code change here? I thought I didn't change the behavior but I could be wrong if `comp` (what we used to use) is more complicated than just forwarding to `torch.testing.allclose`.

Comment (vedaanta-nvidia) [reply thunder/tests/test_cudnn_executor.py:227]: I think its fine as it passes all tests. I can rework tolerances once it starts causing issues.

**Comment by wujingyue** [comment]
> Hey @vedaanta-nvidia, are you interested in reviewing this before I merge?

Hey Vedaanta, do you have more comments before I merge?

File: thunder/executors/cudnnex.py [hunk 3]
```diff
@@ -604,9 +668,20 @@ def _cudnn_sdpa_grad(
...
     )
 
-    g = get_grad(primal)
+    description = """\
+This flag is for enabling nvFuser's zipping optimization that seeks to avoid
+expensive concatenation. https://github.com/NVIDIA/Fuser/issues/1768 has more
+details. When this flag is true, cudnn_sdpa_bwd may cat dQ, dK and dV as one
+tensor and return them as slices of that tensor.
+"""
+    may_cat_grad_qkv: None | bool = get_compile_option("cudnn_sdpa_bwd_may_cat_grad_qkv", description)
```
Comment (vedaanta-nvidia) [inline thunder/executors/cudnnex.py:677]: Nice, I did not know that executors can get hints/kwargs from thunder.compile.

File: thunder/executors/cudnnex.py [hunk 4]
```diff
@@ -464,19 +480,53 @@ def cudnn_sdpa_backward_meta(
...
     *,
     scale: None | float = None,
-) -> (TensorProxy, TensorProxy, TensorProxy):
-    grad_query = TensorProxy(like=query)
-    grad_key = TensorProxy(like=key)
-    grad_value = TensorProxy(like=value)
+    cat_grad_qkv: bool,
+) -> tuple[TensorProxy, ...]:
+    if cat_grad_qkv:
+        grad_qkv = TensorProxy(
```
Comment (vedaanta-nvidia) [inline thunder/executors/cudnnex.py:486]: The compile flag cudnn_sdpa_bwd_may_cat_grad_qkv suggests that cudnnex "may" output concatenated grads. Seems like its just assumed that cudnnex always be able to concatenate grads, as this code path only returns `grad_qkv`.
(Relying on cudnn to be able to do output grads in a strided fashion is a safe assumption for future cudnn versions too, as long as the embedding dim stride is 1.)

File: thunder/executors/cudnnex.py [hunk 5]
```diff
@@ -207,6 +208,11 @@ def compute_NHWC_strides(shape):
 # And when registering for sdpa, cudnn assumes NHWC layout. (See _transform_sdpa_inputs())
 def _sdpa_enforce_input_tensor_contiguity(a: torch.Tensor) -> torch.Tensor:
     if a.stride(-1) == 1:
+        # TODO(vedaanta-nvidia): there's an inconsistency between
```
Comment (vedaanta-nvidia) [inline thunder/executors/cudnnex.py:211]: nice catch. I think `_transform_sdpa_inputs` should not overwrite strides if already set.

So when cudnn checker function is called with TensorProxy:
- assume contiguous stride
- create cudnn execution plan (just fyi, this was removed for the time being in your previous PR)
- store in cache for potential performance boost in execution phase

And then at execution phase with actual torch tensors:
- Use tensor strides to redo cudnn graph/execution plan ([PPOPOSED] DONT OVERWRITE/ASSUME STRIDES)
- If tensor strides are already contiguous, execution benefits from cache hit
- Execute cudnn plan

I think above approach should be safe to do. What do you think?

**Review by vedaanta-nvidia — APPROVED** [review]

File: thunder/executors/cudnnex.py (hunk 5 above)
Comment (wujingyue) [reply thunder/executors/cudnnex.py:211]: > Use tensor strides to redo cudnn graph/execution plan

One potential problem is that the above step can fail even though the checker was happy. Still, with your proposal, cudnnex can at least throw a very clear runtime error when that happens, which is a strict improvement.

That being said, I think the fundamental problem here is: at this moment, a Thunder executor's checkers have no access to strides, but cudnnex needs strides to accurately check an sdpa op can be supported. Am I right? If so, we should probably discuss with Thunder folks on a better, maybe longer-term fix.

File: thunder/executors/cudnnex.py (hunk 4 above)
Comment (wujingyue) [reply thunder/executors/cudnnex.py:486]: > Seems like its just assumed that cudnnex always be able to concatenate grads

No, cat_grad_qkv is not the same as may_cat_grad_qkv. It combines `may_cat_grad_qkv` and [some size/stride checks](https://github.com/Lightning-AI/lightning-thunder/pull/59/files#diff-783705c9be7fba72357ae0678dac1da2961ddb64d37a40eb6e3bf03558f8e313R681) indicating "is able to cat".

**Pull request merged**